metals/src/main/scala/scala/meta/internal/parsing/JavaFoldingRangeExtractor.scala

```diff
@@ -0,0 +1,213 @@
+package scala.meta.internal.parsing
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+import org.eclipse.jdt.core.ToolFactory
+import org.eclipse.jdt.core.compiler.ITerminalSymbols
+import org.eclipse.lsp4j.FoldingRange
+import org.eclipse.lsp4j.FoldingRangeKind
+
+final class JavaFoldingRangeExtractor(
+    text: String,
+    foldOnlyLines: Boolean
+) {
+  private val spanThreshold = 2
+
+  private sealed trait Position
+  private case class LBrace(line: Int, ch: Int) extends Position
+  private case class RBrace(line: Int, ch: Int) extends Position
+
+  private trait Region extends Position {
+    def startLine: Int
+    def startCh: Int
+    def endLine: Int
+    def endCh: Int
+    def kind: String
+  }
+  private case class Comment(
+      startLine: Int,
+      startCh: Int,
+      endLine: Int,
+      endCh: Int
+  ) extends Region {
+    def kind = FoldingRangeKind.Comment
+  }
+
+  private case class Imports(
+      startLine: Int,
+      startCh: Int,
+      endLine: Int,
+      endCh: Int
+  ) extends Region {
+    def kind = FoldingRangeKind.Imports
+  }
+
+  def extract(): List[FoldingRange] = {
+    val scanner = ToolFactory.createScanner(true, true, false, true)
+    scanner.setSource(text.toCharArray())
+
+    @tailrec
+    def swallowUntilSemicolon(token: Int, line: Int): Int = {
+      val addLine = scanner.getCurrentTokenSource.count(_ == '\n')
+      if (token != ITerminalSymbols.TokenNameSEMICOLON) {
+        swallowUntilSemicolon(scanner.getNextToken(), line + addLine)
+      } else {
+        line + addLine
+      }
+    }
+
+    @tailrec
+    def gatherImportLines(
+        token: Int,
+        line: Int,
+        lastImportLine: Int
+    ): (Int, Int, Int, Int) = {
```

Comment:
wdyt about using case class instead of tuple? I know that's it's used only once, but four Ints looks a little bit obscure

Reply:
Done!